src/pool.rs

```diff
@@ -0,0 +1,259 @@
+
+    /// Creates a new object in the pool, reusing storage if possible. This method returns a key
+    /// which can be used to access the storage.
+    ///
+    /// If this function returns `None`, then the shard for the current thread is full and no items
+    /// can be added until some are removed, or the maximum number of shards has been reached.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = Pool::new();
+    /// let value = String::from("Hello");
+    ///
+    /// let key = pool.create_with(value).unwrap();
+    /// assert_eq!(pool.get(key).unwrap(), String::from("Hello"));
+    /// ```
+    pub fn create_with(&self, value: T) -> Option<usize> {
+        let tid = Tid::<C>::current();
+        let mut value = Some(value);
+        test_println!("pool: create_with {:?}", tid);
+        self.shards[tid.as_usize()]
+            .get_initialized_slot(move |item| *item = value)
+            .map(|idx| tid.pack(idx))
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the pool does not contain a value for the given key, `None` is returned instead.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = sharded_slab::Pool::new();
+    /// let mut value = Some(String::from("hello world"));
+    /// let key = pool.create(move |item| *item = value.take().expect("crated twice")).unwrap();
+    ///
+    /// assert_eq!(pool.get(key).unwrap(), String::from("hello world"));
+    /// assert!(pool.get(12345).is_none());
+    /// ```
+    pub fn get(&self, key: usize) -> Option<PoolGuard<'_, T, C>> {
+        let tid = C::unpack_tid(key);
+
+        test_println!("pool: get{:?}; current={:?}", tid, Tid::<C>::current());
+        let inner = self.shards.get(tid.as_usize())?.get(key, |x| x)?;
+
+        Some(PoolGuard {
+            inner,
+            // Safe access as previous line checks for validity
+            shard: &self.shards[tid.as_usize()],
+            key,
+        })
+    }
+
+    /// Remove the value using the storage associated with the given key from the pool, reutrning
+    /// `true` if the value was removed.
+    ///
+    /// Unlike [`clear`], this method does _not_ block the current thread until the value can be
+    /// removed. Instead, if another thread is currently accessing that value, this marks it to be
+    /// removed by that thread when it finishes accessing the value.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = sharded_slab::Pool::new();
+    /// let mut value = Some(String::from("hello world"));
+    /// let key = pool.create(move |item| *item = value.take().expect("crated twice")).unwrap();
+    ///
+    /// assert_eq!(pool.get(key).unwrap(), String::from("hello world"));
+    ///
+    /// pool.clear(key);
+    /// assert!(pool.get(key).is_none());
+    /// ```
```

Comment:
It _might_ also be nice to show an example of how this method's return value can be used to determine if a pooled object was actually cleared. Maybe something like 
```suggestion
    /// ```
    ///
    /// ```rust
    /// # use sharded_slab::Pool;
    /// let pool: Pool<String> = Pool::new();
    /// 
    /// let key = pool.create(|item| item.push_str("hello world")).unwrap();
    ///
    /// // Clearing a key that doesn't exist in the pool will return false.
    /// assert_eq!(pool.clear(key + 69420), false);
    ///
    /// // Clearing a key that does exist in the pool returns true.
    /// assert_eq!(pool.clear(key), true);
    ///
    /// // Clearing a key that has already been cleared will return false.
    /// assert_eq!(pool.clear(key), false);
    /// ```
    ///
```

Reply:
That's a great idea. Also, nice.